src/Graph.cpp

```
@@ -0,0 +1,93 @@
+#include "Graph.hpp"
+
+#include <boost/graph/adjacency_list.hpp>
+#include <unordered_map>
+#include <queue>
+#include <vector>
+#include <string>
+#include <tuple>
+
+namespace booty {
+
+using BoostGraph = boost::adjacency_list<boost::vecS, boost::vecS, boost::undirectedS>;
+using Vertex = boost::graph_traits<BoostGraph>::vertex_descriptor;
+
+class Graph::GraphImpl {
+public:
+    BoostGraph graph;
+    std::unordered_map<std::string, Vertex> vertexMap;
+    std::unordered_map<Vertex, std::string> labelMap;
+};
+
+Graph::Graph() : impl(new GraphImpl) {}
+
+Graph::~Graph() { delete impl; }
+
+void Graph::add_node(const std::string& label) {
+    if (impl->vertexMap.find(label) == impl->vertexMap.end()) {
+        Vertex v = boost::add_vertex(impl->graph);
+        impl->vertexMap[label] = v;
+        impl->labelMap[v] = label;
+    }
+}
+
+void Graph::add_edge(const std::string& from, const std::string& to) {
+    add_node(from);
+    add_node(to);
+    boost::add_edge(impl->vertexMap[from], impl->vertexMap[to], impl->graph);
+}
+
+std::vector<std::string> Graph::bfs(const std::string& startLabel) {
+    std::vector<std::string> result;
+
+    if (impl->vertexMap.find(startLabel) == impl->vertexMap.end())
```

Silently returns an empty list if the start node is missing. Consider signaling this case explicitly to distinguish it from an empty graph, like throwing an exception